scipy/sparse/tests/test_csr.py

```diff
@@ -113,3 +113,25 @@ def test_csr_bool_indexing():
     assert (slice_list2 == slice_array2).all()
     assert (slice_list3 == slice_array3).all()
 
+
+def test_csr_hstack_int64():
+    """
+    Tests if hstack properly promotes to np.int64 if the output sparse matrix
+    has indices too large for np.int64
```

Comment:
Should this be `too large for np.int32` instead of `np.int64`? @rossbar had some concern about excessive memory footprint, but this is mostly a corner case in any case?

@perimosocordiae @hameerabbasi thoughts?

Reply:
@thomasjpfan You need this in scikit-learn, right? And backported as well?

Reply:
Maybe a few quick comments/thoughts on tradeoffs would help--it came up in meeting call today.

Reply:
I'm not sure what the _alternative_ to the expanded memory footprint is while maintaining correctness. In the cases where this PR promotes to `int64` but the implementation on `main` doesn't, we end up with negative columns on `main` due to an integer overflow in the `csr_hstack` function.

Reply:
> In general I think we should specify index dtypes upfront, and error if something doesn't fit.

This would perhaps be a good follow-up PR. I'm thinking that it could be solved by allowing for an `index_dtype` parameter with options `{None, np.int32, np.int64}` where `index_dtype=None` is the default and uses this dynamic resolution?